humility-core/src/arch_rv.rs

```diff
@@ -0,0 +1,322 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::arch::{instr_operands, Register, RegisterField};
+use anyhow::Result;
+use num_traits::ToPrimitive;
+use std::collections::BTreeMap;
+use strum_macros::EnumIter;
+
+#[allow(non_camel_case_types)]
+#[derive(
+    Copy,
+    Clone,
+    Debug,
+    Hash,
+    FromPrimitive,
+    ToPrimitive,
+    PartialEq,
+    Eq,
+    Ord,
+    PartialOrd,
+    EnumIter,
+)]
+
+///
+/// See table 3.3 in the riscv debug spec.
+/// (https://raw.githubusercontent.com/riscv/riscv-debug-spec/master/riscv-debug-stable.pdf#table.3.3)
+/// The pc is read through the DPC register
+///
+/// THOU MUST ORDER THE REGISTERS ACCORDING TO ADDRESS!
+///
+pub enum RVRegister {
+    MSTATUS = 0x300,
+    MISA,
+    MEDELEG,
+    MIDELEG,
+    MIE,
+    MTVEC,
+    MSTATUSH = 0x310,
+    MSCRATCH = 0x340,
+    MEPC,
+    MCAUSE,
+    MTVAL,
+    MIP,
+    PMPCFG0 = 0x3a0,
+    PMPCFG1,
+    PMPCFG2,
+    PMPCFG3,
+    PMPCFG4,
+    PMPCFG5,
+    PMPCFG6,
+    PMPCFG7,
+    PMPCFG8,
+    PMPCFG9,
+    PMPCFG10,
+    PMPCFG11,
+    PMPCFG12,
+    PMPCFG13,
+    PMPCFG14,
+    PMPCFG15,
+    PMPADDR0 = 0x3b0,
+    PMPADDR1,
+    PMPADDR2,
+    PMPADDR3,
+    PMPADDR4,
+    PMPADDR5,
+    PMPADDR6,
+    PMPADDR7,
+    PMPADDR8,
+    PMPADDR9,
+    PMPADDR10,
+    PMPADDR11,
+    PMPADDR12,
+    PMPADDR13,
+    PMPADDR14,
+    PMPADDR15,
+    PMPADDR16,
+    PMPADDR17,
+    PMPADDR18,
+    PMPADDR19,
+    PMPADDR20,
+    PMPADDR21,
+    PMPADDR22,
+    PMPADDR23,
+    PMPADDR24,
+    PMPADDR25,
+    PMPADDR26,
+    PMPADDR27,
+    PMPADDR28,
+    PMPADDR29,
+    PMPADDR30,
+    PMPADDR31,
+    PMPADDR32,
+    PMPADDR33,
+    PMPADDR34,
+    PMPADDR35,
+    PMPADDR36,
+    PMPADDR37,
+    PMPADDR38,
+    PMPADDR39,
+    PMPADDR40,
+    PMPADDR41,
+    PMPADDR42,
+    PMPADDR43,
+    PMPADDR44,
+    PMPADDR45,
+    PMPADDR46,
+    PMPADDR47,
+    PMPADDR48,
+    PMPADDR49,
+    PMPADDR50,
+    PMPADDR51,
+    PMPADDR52,
+    PMPADDR53,
+    PMPADDR54,
+    PMPADDR55,
+    PMPADDR56,
+    PMPADDR57,
+    PMPADDR58,
+    PMPADDR59,
+    PMPADDR60,
+    PMPADDR61,
+    PMPADDR62,
+    PMPADDR63,
+    MSECCFG = 0x747,
+    MSECCFGH = 0x757,
+    DCSR = 0x7b0,
+    PC = 0x7b1,
+    ZERO = 0x1000,
+    RA,
+    SP,
+    GP,
+    TP,
+    T0,
+    T1,
+    T2,
+    S0,
+    S1,
+    A0,
+    A1,
+    A2,
+    A3,
+    A4,
+    A5,
+    A6,
+    A7,
+    S2,
+    S3,
+    S4,
+    S5,
+    S6,
+    S7,
+    S8,
+    S9,
+    S10,
+    S11,
+    T3,
+    T4,
+    T5,
+    T6,
+}
+
+impl RVRegister {
+    pub fn is_general_purpose(&self) -> bool {
+        self.to_u16().unwrap() >= 0x1000
```

Comment:
I think it would be useful to add enums for start and end of the ranges in table 3.3

```
CSR_START = 0,
CSR_END = 0xfff,
GPR_START = 0x1000,
GPR_END = 0x101f,
FPR_START = 0x1020,
FPR_END = 0x103f,
CUSTOM_START = 0xc000,
CUSTOM_END = 0xffff,
```

Those can then be used for checks here rather than magic numbers:
```
pub fn is_general_purpose(&self) -> bool {
        self.to_u16().unwrap() >= GPR_START
                  && self.to_u16().unwrap() <= GPR_END
    }
```

Reply:
Good idea!  I will change this.